### PR TITLE
REPO-2339: Trashcan REST API: optionally restore node to new location

### PR DESCRIPTION
**Pull request by ramunteanu**
- Added option to restore node to a new location

### CODE REVIEW

File: src/test/java/org/alfresco/rest/DeletedNodesTest.java [hunk 1]
```diff
@@ -230,6 +231,41 @@ public void testCreateAndRestore() throws Exception
...
+        nodeTargetAssoc.setTargetParentId(f1Id);
+        nodeTargetAssoc.setAssocType(ASSOC_TYPE_CM_CONTAINS);
+        
+        //restore to new location
+        post(URL_DELETED_NODES+"/"+document.getId()+"/restore", toJsonAsStringNonNull(nodeTargetAssoc), null, 200);
+
+        deleteNode(document.getId());
+        //restore to nonexistent nodeId as the new location
+        nodeTargetAssoc.setTargetParentId("nonexistentTargetNode");
+        post(URL_DELETED_NODES+"/"+document.getId()+"/restore", toJsonAsStringNonNull(nodeTargetAssoc), null, 404);
```
Comment (cristianturlica) [inline src/test/java/org/alfresco/rest/DeletedNodesTest.java]: Code formatting required.

File: src/test/java/org/alfresco/rest/DeletedNodesTest.java [hunk 2]
```diff
@@ -230,6 +231,41 @@ public void testCreateAndRestore() throws Exception
         //Now we can restore it.
         post(URL_DELETED_NODES+"/"+document.getId()+"/restore", null, null, 200);
 
+        deleteNode(document.getId());
+
+        // Create a new nodeTargetAssoc containing target id and association type
+        NodeTargetAssoc nodeTargetAssoc = new NodeTargetAssoc();
+        nodeTargetAssoc.setTargetParentId(f1Id);
+        nodeTargetAssoc.setAssocType(ASSOC_TYPE_CM_CONTAINS);
```
Comment (cristianturlica) [inline src/test/java/org/alfresco/rest/DeletedNodesTest.java]: We should add a test that checks the behaviour when assocType is null.

Comment (ramunteanu) [reply src/test/java/org/alfresco/rest/DeletedNodesTest.java]: Thank you for catching this.
